packages/logging_cloudwatch/amplify_logging_cloudwatch/lib/src/queued_item_store/index_db/indexed_db_adapter.dart

```diff
@@ -0,0 +1,177 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import 'dart:async';
+import 'dart:convert';
+import 'dart:js_util';
+
+import 'package:amplify_core/amplify_core.dart';
+// ignore: implementation_imports
+import 'package:aws_common/src/js/indexed_db.dart';
+import 'package:aws_logging_cloudwatch/aws_logging_cloudwatch.dart';
+import 'package:collection/collection.dart';
+
+// TODO(kylechen): Consider merging/refactoring with existing 'amplify_secure_storage_web - _IndexedDBStorage' class
+/// {@template amplify_logging_cloudwatch_dart.indexed_db_adapter}
+/// Provide indexDB methods for storing/retrieving Strings.
+/// {@endtemplate}
+class IndexedDbAdapter implements QueuedItemStore {
+  /// {@macro amplify_analytics_pinpoint_dart.indexed_db_adapter}
+  IndexedDbAdapter();
+
+  /// The name of the database.
+  ///
+  /// Reference: https://www.w3.org/TR/IndexedDB/#name
+  String get databaseName => 'amplify_logging_cloudwatch';
+
+  /// The name of the object store.
+  ///
+  /// Reference: https://www.w3.org/TR/IndexedDB/#object-store-name
+  final storeName = 'logging_cached_logs';
+
+  late final Future<void> _databaseOpenEvent = _openDatabase();
+
+  late final IDBDatabase _database;
+
+  int _currentTotalByteSize = 0;
+
+  /// Checks for IDB availability and attempts to open the database.
+  Future<void> _openDatabase() async {
+    final db = indexedDB;
+    if (db == null) {
+      throw const InvalidStateException('IndexedDB is not available');
+    }
+    final openRequest = db.open(databaseName, 1)
+      ..onupgradeneeded = (event) {
+        final database = event.target.result;
+        final objectStoreNames = database.objectStoreNames;
+        if (!objectStoreNames.contains(storeName)) {
+          database.createObjectStore(
+            storeName,
+            keyPath: 'id',
+            autoIncrement: true,
+          );
+        }
+      };
+    _database = await openRequest.future;
+  }
+
+  /// Returns a new [IDBObjectStore] instance after waiting for initialization
+  /// to complete.
+  IDBObjectStore _getObjectStore() {
+    final transaction = _database.transaction(
+      storeName,
+      mode: IDBTransactionMode.readwrite,
+    );
+    final store = transaction.objectStore(storeName);
+    return store;
+  }
+
+  @override
+  Future<void> addItem(String string, String timestamp) async {
+    _currentTotalByteSize += utf8.encode(string).length;
+    _currentTotalByteSize += utf8.encode(timestamp).length;
+    _currentTotalByteSize += 8; // 8 bytes for the id
```

Comment:
This logic is duplicated in multiple places. Best to consolidate in it a helper method somewhere.